cmd/batch-handlers.go

```diff
@@ -0,0 +1,1068 @@
+	if err != nil {
+		return err
+	}
+
+	return saveConfig(ctx, api, pathJoin(jobLocation, batchReplName), buf)
+}
+
+func (ri *batchJobInfo) countItem(size int64, success bool) {
+	if ri == nil {
+		return
+	}
+	if success {
+		ri.Objects++
+		ri.BytesTransferred += size
+	} else {
+		ri.ObjectsFailed++
+		ri.BytesFailed += size
+	}
+}
+
+func (ri *batchJobInfo) updateAfter(ctx context.Context, api ObjectLayer, duration time.Duration, jobLocation string) error {
+	if ri == nil {
+		return errInvalidArgument
+	}
+	now := UTCNow()
+	if now.Sub(ri.LastUpdate) >= duration {
+		if serverDebugLog {
+			console.Debugf("batchReplicate: persisting batchReplication info on drive: threshold:%s, batchRepl:%#v\n", now.Sub(ri.LastUpdate), ri)
+		}
+		ri.LastUpdate = now
+		ri.Version = batchReplVersionV1
+		return ri.save(ctx, api, jobLocation)
+	}
+	return nil
+}
+
+func (ri *batchJobInfo) trackCurrentBucketObject(bucket string, info ObjectInfo, failed bool) {
+	if ri == nil {
+		return
+	}
+	ri.Bucket = bucket
+	ri.Object = info.Name
+	ri.countItem(info.Size, failed)
+}
+
+// Start start the batch replication job, resumes if there was a pending job via "job.ID"
+func (r *BatchJobReplicateV1) Start(ctx context.Context, api ObjectLayer, job BatchJobRequest) error {
+	ri := &batchJobInfo{
+		JobID:     job.ID,
+		JobType:   string(job.Type()),
+		StartTime: job.Started,
+	}
+	if err := ri.load(ctx, api, job); err != nil {
+		return err
+	}
+	globalBatchJobsMetrics.save(job.ID, ri.clone())
+	lastObject := ri.Object
+
+	delay := job.Replicate.Flags.Retry.Delay
+	if delay == 0 {
+		delay = batchReplJobDefaultRetryDelay
+	}
+	rnd := rand.New(rand.NewSource(time.Now().UnixNano()))
+
+	skip := func(info FileInfo) (ok bool) {
+		if r.Flags.Filter.OlderThan > 0 && time.Since(info.ModTime) < r.Flags.Filter.OlderThan {
+			// skip all objects that are newer than specified older duration
+			return false
+		}
+
+		if r.Flags.Filter.NewerThan > 0 && time.Since(info.ModTime) >= r.Flags.Filter.NewerThan {
+			// skip all objects that are older than specified newer duration
+			return false
+		}
+
+		if !r.Flags.Filter.CreatedAfter.IsZero() && r.Flags.Filter.CreatedAfter.Before(info.ModTime) {
+			// skip all objects that are created before the specified time.
+			return false
+		}
+
+		if !r.Flags.Filter.CreatedBefore.IsZero() && r.Flags.Filter.CreatedBefore.After(info.ModTime) {
+			// skip all objects that are created after the specified time.
+			return false
+		}
+
+		return true
+	}
+
+	u, err := url.Parse(r.Target.Endpoint)
+	if err != nil {
+		return err
+	}
+
+	cred := r.Target.Creds
+
+	c, err := miniogo.NewCore(u.Host, &miniogo.Options{
+		Creds:     credentials.NewStaticV4(cred.AccessKey, cred.SecretKey, cred.SessionToken),
+		Secure:    u.Scheme == "https",
+		Transport: getRemoteInstanceTransport,
+	})
+	if err != nil {
+		return err
+	}
+
+	retryAttempts := ri.RetryAttempts
+	for attempts := 1; attempts <= retryAttempts; attempts++ {
```

Comment:
not sure why we want to do retries on the entire object walk here - may be better to do it at object level to avoid copying entire list over on each re-try. Might be even better to just save failures MRF style and reattempt for configured retry count after this walk is done

Reply:
We can do that, this part is not optimized.